akka-actor-tests/src/test/scala/akka/actor/CoordinatedShutdownSpec.scala

```diff
@@ -320,18 +322,61 @@ class CoordinatedShutdownSpec
       confWithOverrides.getInt("exit-code") should ===(-1)
     }
 
-    // this must be the last test, since it terminates the ActorSystem
     "terminate ActorSystem" in {
-      Await.result(CoordinatedShutdown(system).run(CustomReason), 10.seconds) should ===(Done)
-      system.whenTerminated.isCompleted should ===(true)
-      CoordinatedShutdown(system).shutdownReason() === (Some(CustomReason))
+      val sys = ActorSystem(system.name, system.settings.config)
+      try {
+        Await.result(CoordinatedShutdown(sys).run(CustomReason), 10.seconds) should ===(Done)
+        sys.whenTerminated.isCompleted should ===(true)
+        CoordinatedShutdown(sys).shutdownReason() should ===(Some(CustomReason))
+      } finally {
+        shutdown(sys)
+      }
+    }
+
+    "be run by ActorSystem.terminate" in {
+      val sys = ActorSystem(system.name, system.settings.config)
+      try {
+        Await.result(sys.terminate(), 10.seconds)
+        sys.whenTerminated.isCompleted should ===(true)
+        CoordinatedShutdown(sys).shutdownReason() should ===(Some(CoordinatedShutdown.ActorSystemTerminateReason))
+      } finally {
+        shutdown(sys)
+      }
+    }
+
+    "not be run by ActorSystem.terminate when run-by-actor-system-terminate=off" in {
+      val sys = ActorSystem(
+        system.name,
+        ConfigFactory
+          .parseString("akka.coordinated-shutdown.run-by-actor-system-terminate = off")
+          .withFallback(system.settings.config))
+      try {
+        Await.result(sys.terminate(), 10.seconds)
+        sys.whenTerminated.isCompleted should ===(true)
+        CoordinatedShutdown(sys).shutdownReason() should ===(None)
+      } finally {
+        shutdown(sys)
+      }
+    }
+
+    "not allow terminate-actor-system=off && run-by-actor-system-terminate=on" in {
+      intercept[ConfigurationException] {
+        val sys = ActorSystem(
+          system.name,
+          ConfigFactory
+            .parseString("akka.coordinated-shutdown.terminate-actor-system = off")
+            .withFallback(system.settings.config))
+        // will only get here if test is failing
+        shutdown(sys)
```

Comment:
Doesn't this test leak an `ActorSystem`?

Reply:
it's testing that a `ConfigurationException` is thrown from `ActorSystem.apply` (constructor) so normally it's not even created. this shutdown is only for the case that there is a test failure, i.e. when exception not thrown